proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,369 @@
+
+## `variant.lift #typeidx $tag: #u32`
+
+The type of this instruction is:
+```
+[$payload] -> [$type]
+
+Where:
+  * $type is a type of the form (variant $options) taken from #typeidx
+  * $tag is an index of an (option #interface-valtype) from `$options`
+  * $payload = #interface-valtype
+```
+
+The output is a variant interface value with `tag` given by `$tag` and `payload` given by input.
+
+## `variant.lower_tags`
+
+The full syntax of this instruction is:
+```
+variant.lower_tag #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [i32]
+
+Where:
+ * $type is a valid type of the form (variant ..) taken from #typeidx
+```
+
+The output is the value of the `tag` of the input `variant`.
+
+## `variant.lower`
+
+This is a structured instruction with the following syntax:
+```
+variant.lower #typeidx $result: #interface-valtype*
+  (case $tag #interface-instr*)+
+end
+```
+
+The type of this instruction is:
+```
+[$type] -> [$result]
+
+Where:
+  * $type is a valid type of the form (variant $options) taken from #typeidx
+  * Each $tag uniquely matches an index of an option of $options
+  * Every option of $options is referenced by a case
+  * Each #interface-instr* is valid with the type [] -> [$result]
+```
+
+The output is the value of executed the matching `case` for the `tag` of the input `variant`.
+
+## `array.lift_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lift_memory #typeidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$base: i32, $length: i32] -> [$type]
+
+Where:
+  * $type is a type of the form (array $inner) taken from #typeidx
+  * #interface-instr* is valid with type [i32] -> [$inner]
+```
+
+This instruction will execute the following steps:
+```
+  1. Let $offsets = take([$base, $base + $width, $base + $width * 2, ..], $length)
+  2. Map #interface-instr* upon $offsets to yield the elements of the array
+```
+
+## `array.lower_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lower_memory #typeidx #memidx $alloc: #funcidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$type] -> []
+
+Where:
+  * $type is a valid type of the form (array $inner) taken from #typeidx
+  * $alloc has a valid functype of [i32] -> [i32]
+  * #interface-instr* is valid with type [i32, $inner] -> []
+```
+
+This instruction will execute the following steps:
+```
+  1. Let $elems be the elements of the array
+  2. Let $length be the length of the array
+  3. Let $bytes be '$length * $width'
+  4. Let $base = $alloc($bytes)
+  5. Let $offsets = take([$base, $base + $width, $base + $width * 2, ..], $length)
+  6. Let $params = zip($offsets, $elems)
+  7. Map #interface-instr* upon $params
+```
+
+## `callback.lift`
+
+The full syntax of this instruction is:
+```
+callback.lift #typeidx $environment: #interface-valtype* $exec: #interface-funcidx $dtor: #interface-funcidx
```

Comment:
Maybe lets hold off on adding callback values for a bit.

Reply:
we do not need a special instruction for callbacks. 
We DO need function reference types (which includes a func.bind instruction)

Reply:
Okay, I've dropped callback types.